File: equater-android/app/src/main/java/com/equater/equater/transaction/TransactionDetailCard.kt
package com.equater.equater.transaction

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material.Divider
import androidx.compose.material.MaterialTheme
import androidx.compose.material.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.unit.dp
import com.equater.equater.extensions.formatMonthDayYear
import com.equater.equater.extensions.toLocalDateTime
import com.equater.equater.sharedExpenseCreation.TransactionStory
import com.equater.equater.ui.backgroundSecondary
import com.equater.equater.ui.textSecondaryColor

@Composable fun TransactionDetailCard(story: TransactionStory) {
    fun getDateDescription(): String {
        val transaction = story.transaction

        if (transaction.dateTimeTransferredToDestination != null) {
            return "Paid On"
        }

        if (transaction.dateTimeTransactionScheduled != null) {
            return "Scheduled For"
        }

        return "Initiated On"
    }

    fun getDate(): String {
        val transaction = story.transaction
        val date = transaction.dateTimeTransferredToDestination
            ?: transaction.dateTimeTransactionScheduled
            ?: transaction.dateTimeInitiated

        return date.toLocalDateTime().formatMonthDayYear()
    }

    fun getStatus(): String {
        val status = story.transaction.dwollaStatus?.value ?: "pending"

        return status.replaceFirstChar { it.uppercase() }
    }

    Column(
        modifier = Modifier
            .clip(RoundedCornerShape(8.dp))
            .background(backgroundSecondary())
            .padding(vertical = 12.dp, horizontal = 16.dp)
    ) {
        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(vertical = 4.dp),
            horizontalArrangement = Arrangement.SpaceBetween
        ) {
            Text(text = getDateDescription(), style = MaterialTheme.typography.body1)
            Text(text = getDate(), style = MaterialTheme.typography.body1)
        }

        Divider(color = textSecondaryColor().copy(alpha = 0.5f))

        Row(
            modifier = Modifier
                .fillMaxWidth()
                .padding(vertical = 4.dp),
            horizontalArrangement = Arrangement.SpaceBetween
        ) {
            Text(text = "Status", style = MaterialTheme.typography.body1)
            Text(text = getStatus(), style = MaterialTheme.typography.body1)
        }
    }
}

File: equater-ios/Equater/Modules/Transaction/TransactionDetailCard.swift
//
//  TransactionDetailCard.swift
//  Equater
//
//  Created by Robert B. Menke on 7/5/20.
//  Copyright © 2020 beauchampsullivan. All rights reserved.
//

import SwiftUI

struct TransactionDetailCard: View {
	let story: TransactionStory

	var body: some View {
		Card {
			VStack(spacing: 4) {
				HStack {
					AppText(self.getDateDescription(), font: .primaryText)
					Spacer()
					AppText(self.getDate(), font: .subText)
				}
				.frameFillWidth(height: 20, alignment: .bottom)
				.padding([.leading, .trailing], 16)

				Divider()
					.frame(height: 1)
					.background(AppColor.backgroundSecondary.color)
					.padding([.leading, .trailing], 8)

				HStack {
					AppText("Status", font: .primaryText)
					Spacer()
					AppText(self.getStatus(), font: .subText)
				}
				.frameFillWidth(height: 20, alignment: .top)
				.padding([.leading, .trailing], 16)
			}
		}
	}

	private func getDateDescription() -> String {
		let transaction = story.transaction

		if transaction.dateTimeTransferredToDestination != nil {
			return "Paid On"
		}

		if transaction.dateTimeTransactionScheduled != nil {
			return "Scheduled For"
		}

		return "Initiated On"
	}

	private func getDate() -> String {
		let transaction = story.transaction
		let date = transaction.dateTimeTransferredToDestination ?? transaction.dateTimeTransactionScheduled ?? transaction.dateTimeInitiated

		guard let isoDate = date.toISODate() else { return "??" }

		return isoDate.formatMonthDayYear()
	}

	private func getStatus() -> String {
		let status = story.transaction.dwollaStatus?.rawValue ?? "pending"

		return status.capitalizingFirstLetter()
	}
}

struct TransactionDetailCard_Previews: PreviewProvider {
	static var previews: some View {
		TransactionDetailCard(story: transactionStoryFake)
	}
}
